File: app/src/main/java/com/lunesu/pengchauferry/DbOpenHelper.kt
package com.lunesu.pengchauferry

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

class DbOpenHelper(context: Context?): SQLiteOpenHelper(context, if (context == null) null else DB_NAME, null, DB_VERSION) {

    companion object {
        const val TIMES = "times"
        const val HOLIDAYS = "holidays"

        private const val DB_NAME = "ferry2"
        private const val DB_VERSION = 9

        private const val CREATE_TABLE_TIMES =
            "CREATE TABLE $TIMES (time TEXT NOT NULL, `from` TEXT NOT NULL, `to` TEXT NOT NULL, durationMin INTEGER NOT NULL, days INTEGER NOT NULL, fare TEXT);"
        private const val CREATE_INDEX_TIMES = "CREATE INDEX times_from_time ON $TIMES (`from`,time);"
        private const val CREATE_UNIQUE_INDEX_TIMES = "CREATE UNIQUE INDEX times_unique ON $TIMES (time,`from`,`to`,days);"

        private const val CREATE_TABLE_HOLIDAYS =
            "CREATE TABLE $HOLIDAYS (date TEXT NOT NULL);"
        private const val CREATE_INDEX_HOLIDAYS = "CREATE UNIQUE INDEX holidays_date ON $HOLIDAYS (date);"
    }

    override fun onCreate(db: SQLiteDatabase?) {
        db?.execSQL(CREATE_TABLE_TIMES)
        db?.execSQL(CREATE_INDEX_TIMES)
        db?.execSQL(CREATE_UNIQUE_INDEX_TIMES)
        db?.execSQL(CREATE_TABLE_HOLIDAYS)
        db?.execSQL(CREATE_INDEX_HOLIDAYS)
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        when (oldVersion) {
            3 -> db?.execSQL(CREATE_TABLE_HOLIDAYS)
            4 -> db?.execSQL(CREATE_INDEX_HOLIDAYS)
            6 -> db?.execSQL(CREATE_UNIQUE_INDEX_TIMES)
            7 -> {
                db?.execSQL("DROP INDEX times_from;")
                db?.execSQL(CREATE_INDEX_TIMES)
            }
            8 -> db?.execSQL("ALTER TABLE $TIMES ADD COLUMN fare TEXT;")
        }
    }

}

File: iOS/PengChauFerry/DbOpenHelper.swift
//
//  DbOpenHelper.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 23/2/2019.
//  Copyright © 2019 Lionello Lunesu. All rights reserved.
//

import Foundation
import SQLite3

class DBOpenHelper: SQLiteOpenHelper {
    init() {
        super.init(name: "ferry", version: DBOpenHelper.DATABASE_VERSION)
    }

    static let TIMES = "times"
    static let HOLIDAYS = "holidays"

    static private let DATABASE_VERSION = 7
    static private let CREATE_TABLE_TIMES =
        "CREATE TABLE \(TIMES) (time TEXT NOT NULL, `from` TEXT NOT NULL, `to` TEXT NOT NULL, durationMin INTEGER NOT NULL, days INTEGER NOT NULL, via TEXT);"
    static private let CREATE_INDEX_TIMES = "CREATE INDEX times_from_time ON \(TIMES) (`from`,time)"
    static private let CREATE_UNIQUE_INDEX_TIMES =
        "CREATE UNIQUE INDEX times_unique ON \(TIMES) (time,`from`,`to`,days)"
    static private let CREATE_TABLE_HOLIDAYS = "CREATE TABLE \(HOLIDAYS) (date TEXT NOT NULL);"
    static private let CREATE_INDEX_HOLIDAYS = "CREATE UNIQUE INDEX holidays_date ON \(HOLIDAYS) (date)"

    override func onCreate(_ db: SQLiteDatabase?) throws {
        try db?.execSQL(DBOpenHelper.CREATE_TABLE_TIMES)
        try db?.execSQL(DBOpenHelper.CREATE_INDEX_TIMES)
        try db?.execSQL(DBOpenHelper.CREATE_UNIQUE_INDEX_TIMES)
        try db?.execSQL(DBOpenHelper.CREATE_TABLE_HOLIDAYS)
        try db?.execSQL(DBOpenHelper.CREATE_INDEX_HOLIDAYS)
    }

    override func onUpgrade(_ database: SQLiteDatabase?, _ oldVersion: Int, _ newVersion: Int) throws {
        // todo
    }

}
